searx/network/context.py

```diff
@@ -0,0 +1,387 @@
+
+    def _get_new_client_from_factory(self):
+        return _RetryFunctionHTTPClient(super()._get_new_client_from_factory(), self)
+
+
+class _RetryFunctionHTTPClient(ABCHTTPClient):
+    """Companion class of NetworkContextRetryFunction
+
+    Do one thing: if the retries count of the NetworkContext is zero and there is a SoftRetryHTTPException,
+    then the send method catch this exception and returns the HTTP response.
+    This make sure the SoftRetryHTTPException exception is not seen outside the searx.network module.
+    """
+
+    def __init__(self, http_client: ABCHTTPClient, network_context: NetworkContextRetryFunction):
+        self.http_client = http_client
+        self.network_context = network_context
+
+    def send(self, stream: bool, method: str, url: str, **kwargs) -> httpx.Response:
+        try:
+            return self.http_client.send(stream, method, url, **kwargs)
+        except SoftRetryHTTPException as e:
+            if self.network_context._retries <= 0:  # pylint: disable=protected-access
+                return e.response
+            raise e
+
+    def close(self):
+        return self.http_client.close()
+
+    @property
+    def is_closed(self) -> bool:
+        return self.http_client.is_closed
+
+
+## NetworkContextRetrySameHTTPClient
+
+
+class NetworkContextRetrySameHTTPClient(NetworkContext):
+    """When an HTTP request fails, this NetworkContext tries again
+    the same HTTP request with the same HTTP client
+
+    The implementation wraps the provided ABCHTTPClient with _RetrySameHTTPClient."""
+
+    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        try:
+            self._set_http_client()
+            return func(*args, **kwargs)  # type: ignore
+        finally:
+            self._reset_http_client()
+
+    def _get_new_client_from_factory(self):
+        return _RetrySameHTTPClient(super()._get_new_client_from_factory(), self)
+
+
+class _RetrySameHTTPClient(ABCHTTPClient):
+    """Companion class of NetworkContextRetrySameHTTPClient"""
+
+    def __init__(self, http_client: ABCHTTPClient, network_content: NetworkContextRetrySameHTTPClient):
+        self.http_client = http_client
+        self.network_content = network_content
+
+    def send(self, stream: bool, method: str, url: str, **kwargs) -> httpx.Response:
+        retries = self.network_content._retries  # pylint: disable=protected-access
+        while retries >= 0 and self.network_content.get_remaining_time() > 0:
+            try:
+                return self.http_client.send(stream, method, url, **kwargs)
+            except SoftRetryHTTPException as e:
+                if retries <= 0:
+                    return e.response
+            except (ssl.SSLError, httpx.RequestError, httpx.HTTPStatusError) as e:
+                if retries <= 0:
+                    raise e
+            retries -= 1
+        return self.http_client.send(stream, method, url, **kwargs)
```

Comment:
These lines are used about three times with very small modifications only for the `Retry...HTTPClient`'s, I wonder if we can somehow clean that up using a function or make the `Retry...HTTPClient`'s inherit from each other?